### PR TITLE
Implement semihosting debug output redirection to usb-uart port.

### PR DESCRIPTION
**Pull request by sullin**
Some IDEs (ie Eclipse) do not show the semihosting output via gdb. In addition, it is sometimes more comfortable to use a serial port terminal program (like when debugging to UART directly but for the boards where UART is not available). For such use cases, it would be useful to redirect the semihost stdout and stderr streams to the already existing usb-uart port (that is probably unused when semihosting is used).
This patch adds a configuration command "redirect_stdout". By setting "monitor redirect_stdout enable", the semihosting stdout and stderr are sent to usb-uart channel.
Tested on stlink platform.

### CODE REVIEW

**Comment by UweBonnes** [comment]
Is your patch still needed with Koen De Vleeschauwer recent work on semihosting and tracing?

**Comment by sullin** [comment]
> Is your patch still needed with Koen De Vleeschauwer recent work on semihosting and tracing?

I haven't been up to date with the upstream, sorry. I used that for a project that did not have debug UART and that is done for now.

**Comment by UweBonnes** [comment]
Please recheck against current BMPO with many semihostinh enhancements!

**Comment by UweBonnes** [comment]
Please recheck against current BMP with many semihosting enhancements!

File: src/target/target.c [hunk 1]
```diff
@@ -451,6 +465,18 @@ int target_command(target *t, int argc, const char *argv[])
 	return -1;
 }
 
+#ifdef PLATFORM_HAS_USBUART
+static bool cmd_redirect_stdout(target *t, int argc, const char **argv)
+{
+	if (argc == 1)
+		tc_printf(t, "Semihosting stdout redirection: %s\n",
```
Comment (dragonmux) [inline src/target/target.c:472]: Should use gdb_outf here.

**Review by dragonmux — COMMENTED** [review]
These changes generally look ok, though they need rebasing and fixing against the current state of the repo and we'd like to see some changes in how the extra command is added - but there's nothing really objectionable here and the stuff Uwe was pointing at is a red herring as it can't do what this provides.

File: src/target/target.c [hunk 2]
```diff
@@ -21,11 +21,23 @@
...
+#include <unistd.h>
 
 target *target_list = NULL;
 
+#define STDOUT_READ_BUF_SIZE	64
+
+#ifdef PLATFORM_HAS_USBUART
+static bool cmd_redirect_stdout(target *t, int argc, const char **argv);
+const struct command_s semihost_cmd_list[] = {
+	{"redirect_stdout", (cmd_handler)cmd_redirect_stdout, "Redirect semihosting stdout to USB UART"},
```
[the remark this answers is not in the document]
Comment (dragonmux) [reply src/target/target.c:36]: Correction, this should be in cortextm.c

**Comment by dragonmux** [comment]
Given this appears to be a dead PR, we have rescued the contribution to a feature branch where we have rebased it and addressed the review points above. A new PR that supersedes this one has been opened as #.

Sorry to see this grind to a halt because it wasn't addressed in a timely manner, and thank you for the contribution. It is appreciated and we has preserved authorship in the new PR so you will be credited for the work in the release notes.

**Pull request closed without merging**